### ISSUE TITLE
Error message on boot "no IP address found for ovpnc3:network"

### ISSUE THREAD

**Issue by kingfisher77** [opnsense/core#2171]
Since update from 17.x to 18.1.1 booting opnsense throws an error message (see angle brackets
below). The involved network ovpnc3:network is still working after boot.

Everything was pretty fine before upgrade. What do we miss? 

```
2018-02-05T23:00:12+01:00 fw.lan.example.com devd: Popping table
2018-02-05T23:00:12+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:12+01:00 fw.lan.example.com kernel: Configuring WAN interface...
2018-02-05T23:00:12+01:00 fw.lan.example.com kernel: re2: link state changed to UP
2018-02-05T23:00:12+01:00 fw.lan.example.com devd: Pushing table
2018-02-05T23:00:12+01:00 fw.lan.example.com devd: Processing notify event
2018-02-05T23:00:12+01:00 fw.lan.example.com devd: Testing media type of re2 against 0x20
2018-02-05T23:00:12+01:00 fw.lan.example.com devd: re2 has media type 0x20
2018-02-05T23:00:12+01:00 fw.lan.example.com devd: Executing '/usr/local/opnsense/service/configd_ctl.py interface linkup start re2'
2018-02-05T23:00:12+01:00 fw.lan.example.com configd.py: [08968d87-5323-4d0a-937b-97f804b28e4a] Linkup starting re2
2018-02-05T23:00:12+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:12+01:00 fw.lan.example.com kernel: Configuring WLAN_GUEST interface...
2018-02-05T23:00:13+01:00 fw.lan.example.com devd: Popping table
2018-02-05T23:00:13+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:13+01:00 fw.lan.example.com kernel: Configuring WLAN_PUBLIC interface...
2018-02-05T23:00:13+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:13+01:00 fw.lan.example.com kernel: Configuring WLAN_SERVICE interface...
2018-02-05T23:00:13+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:13+01:00 fw.lan.example.com kernel: Configuring VPN_CLIENT_SHIFT interface...
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: Configuring VPN_SERVER_WALD interface...
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: done.
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: Generating /etc/resolv.conf...done.
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: Configuring firewall.
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: pflog0: promiscuous mode enabled
2018-02-05T23:00:14+01:00 fw.lan.example.com kernel: .
2018-02-05T23:00:15+01:00 fw.lan.example.com kernel: ..
<<< 3 Minutes to here
2018-02-05T23:03:17+01:00 fw.lan.example.com opnsense: /usr/local/etc/rc.bootup: New alert found: There were error(s) loading the rules: no IP address found for ovpnc3:network - The line in question reads [0]:
>>>
2018-02-05T23:03:17+01:00 fw.lan.example.com kernel:
2018-02-05T23:03:17+01:00 fw.lan.example.com kernel: Warning: A non-numeric value encountered in /usr/local/etc/inc/filter.inc on line 473
2018-02-05T23:03:17+01:00 fw.lan.example.com kernel: Setting up gateway monitors...
2018-02-05T23:03:17+01:00 fw.lan.example.com apinger: Starting Alarm Pinger, apinger(59765)
2018-02-05T23:03:17+01:00 fw.lan.example.com kernel: done.
```

**Comment by fichtner** [comment]
Will you find the lines using "ovpnc3:network" in /tmp/rules.debug and paste them here?


Thanks,
Franco

**Comment by kingfisher77** [comment]
Here it is

`pass out route-to ( ovpnc3 10.99.0.1 ) from {ovpnc3} to {!(ovpnc3:network)} keep state allow-opts label "let out anything from firewall host itself"`

**Comment by fichtner** [comment]
quick fix may be: Firewall: Settings: Advanced: set "Disable automatic rules which force local services to use the assigned interface gateway"

Not sure about a proper fix yet, CC @AdSchellevis

**Comment by fichtner** [comment]
Is that a tun or tap OpenVPN client? tap?

**Comment by kingfisher77** [comment]
Its `tun`

**Comment by kingfisher77** [comment]
There is also a "Problem detected" in the GUI with the option to send it (which i did). There is a PHP Error mentioned:
`
[05-Feb-2018 23:19:25 Europe/Berlin] PHP Warning:  A non-numeric value encountered in /usr/local/etc/inc/filter.inc on line 473
` 
There is this:

```
        $line_error = explode(":", $rules_error[0]);
        if (isset($line_error[1])) {
            // only report issues with line numbers
            $line_number = $line_error[1];
            $line_split = file('/tmp/rules.debug');
            if (is_array($line_split)) {
                $config_line_error = sprintf(' - ' . gettext('The line in question reads [%d]: %s'), $line_number, $line_split[$line_number-1]);
            }
        }
```

**Comment by fichtner** [comment]
all due to the same error: there is no viable address on your tun client. Not sure why yet.

**Comment by AdSchellevis** [comment]
ok, this is odd, it should check for address type using https://github.com/opnsense/core/blob/18.1/src/etc/inc/filter.lib.inc#L535-L538, which seems to be the same in 17.7

@kingfisher77 what happens if you try to load the rules manually after boot? (pfctl -f /tmp/rules.debug)

**Comment by kingfisher77** [comment]
Looks good:

```
root@fw:~ # pfctl -f /tmp/rules.debug
root@fw:~ #
```

**Comment by kingfisher77** [comment]
Investigated a bit more and tried things back and forth without success. It seems this is a bug somewhere in the software. The OpenVPN client works after boot but it is nit nice to have this error on startup. How could i support narrowing down the error?

**Comment by AdSchellevis** [comment]
does this always happen on reboot?

**Comment by kingfisher77** [comment]
Yes, always. Even if I switch off the interface manually via Web GUI and turn it on again, the error occurs.

**Comment by kingfisher77** [comment]
Today i removed the OpenVPN client where the error at boot up occured. Now the "next in line" is producing the error, the OpenVPN Server:

```
2018-02-18T10:08:43+01:00 fw.lan.mydomain.net opnsense: /usr/local/etc/rc.bootup: New alert found: There were error(s) loading the rules: no IP address found for ovpns1:network - The line in ques
tion reads [0]:
```

**Comment by kingfisher77** [comment]
Added the server and the client from the scratch - while activating the interface the errors appear again.

`There were error(s) loading the rules: no IP address found for ovpns1:network - The line in question reads [0]:` 

`There were error(s) loading the rules: no IP address found for ovpnc2:network - The line in question reads [0]:`

**Comment by AdSchellevis** [comment]
Could it be that you added the OpenVPN interface using Interfaces->Assignments? Normally ovpns1 shouldn't be included in the automatically generated rules. Easiest option in your case is probably to disable "Disable force gateway" in the firewall settings, test if that actually solves the issue (which is what I expect) and add the manual rules where applicable if you use multiple gateways.

**Comment by kingfisher77** [comment]
On 21 Feb 2018, at 19:19, Ad Schellevis wrote:

> Could it be that you added the OpenVPN interface using 
> Interfaces->Assignments? Normally ovpns1 shouldn't be included in the 
> automatically generated rules. Easiest option in your case is probably 
> to disable "Disable force gateway" in the firewall settings, test if 
> that actually solves the issue (which is what I expect) and add the 
> manual rules where applicable if you use multiple gateways.

Yes we have had assignments for the OpenVPN interfaces. Since update to 
18.x the error appeared.

"Disable force gateway" does not lead to an result.

We removed the assignments and the error disappeared (logically). But we 
need an interface für the VPN client to configure Firewall>NAT>Outbound 
properly for natting VPN-Server clients to the site-to-site VPN. For now 
we assigned a fixed address to the interface of the VPN client. That 
prevents the error also.

Not sure if this is the smartest fix.

Also not sure how the system under the hood works - why is there 
„OpenVPN“ under Firewall>Rules when we create a VPN. Do all VPN 
share this vehicle to adjust firewall rules to an vpn network?

Curious,

Tobias


-- 
Tobias Kirchhofer
tobias@kirchhofer.net

**Comment by AdSchellevis** [comment]
openvpn is an interface group, containing all openvpn interfaces (server and client). For most scenarios the group is enough to manage all connections.

It's odd if "Disable force gateway" doesn't resolve the issue, the rule should have come from there. Can you check your rules after a reboot when it's disabled?

**Comment by kingfisher77** [comment]
On 22 Feb 2018, at 13:56, Ad Schellevis wrote:

> openvpn is an interface group, containing all openvpn interfaces 
> (server and client). For most scenarios the group is enough to manage 
> all connections.
>
> It's odd if "Disable force gateway" doesn't resolve the issue, the 
> rule should have come from there. Can you check your rules after a 
> reboot when it's disabled?

How could i do the check? In a file per cli? What should i check 
exactly?


-- 
Tobias Kirchhofer
tobias@kirchhofer.net

**Comment by AdSchellevis** [comment]
In a console indeed:

`grep ovpn /tmp/rules.debug
`

**Comment by kingfisher77** [comment]
Our setup in terms of interfaces:
- one VPN client with an assigned interface (ovpnc2)
- one VPN server without an assigned interface
- three physical interfaces
- four VLAN interfaces

Clients of the VPN server can evenly access all local nets and the networks accessible through the  VPN client.

Rules with „Disable force gateway"
```
root@fw:~ # grep ovpn /tmp/rules.debug
scrub on ovpnc2 all
nat on ovpnc2 inet from $NET_VPN_SERVER to any -> ovpnc2 port 1024:65535
nat on ovpnc2 inet from $HOST_SHIFTVPN_ALLOWED to any -> ovpnc2 port 1024:65535
antispoof log for ovpnc2
# block in quick on ovpnc2 from {<bogons>} to {any} label "block bogon IPv4 networks from VPN_CLIENT_SHIFT"
# block in quick on ovpnc2 from {<bogonsv6>} to {any} label "block bogon IPv6 networks from VPN_CLIENT_SHIFT"
# block in quick on ovpnc2 from {10.0.0.0/8,127.0.0.0/8,100.64.0.0/10,172.16.0.0/12,192.168.0.0/16,fc00::/7} to {any} label "Block private networks from VPN_CLIENT_SHIFT"
```

Rules without „Disable force gateway"
```
root@fw:~ # grep ovpn /tmp/rules.debug
scrub on ovpnc2 all
nat on ovpnc2 inet from $NET_VPN_SERVER to any -> ovpnc2 port 1024:65535
nat on ovpnc2 inet from $HOST_SHIFTVPN_ALLOWED to any -> ovpnc2 port 1024:65535
antispoof log for ovpnc2
# block in quick on ovpnc2 from {<bogons>} to {any} label "block bogon IPv4 networks from VPN_CLIENT_SHIFT"
# block in quick on ovpnc2 from {<bogonsv6>} to {any} label "block bogon IPv6 networks from VPN_CLIENT_SHIFT"
# block in quick on ovpnc2 from {10.0.0.0/8,127.0.0.0/8,100.64.0.0/10,172.16.0.0/12,192.168.0.0/16,fc00::/7} to {any} label "Block private networks from VPN_CLIENT_SHIFT"
pass out route-to ( ovpnc2 10.99.0.1 ) from {ovpnc2} to {!(ovpnc2:network)} keep state allow-opts label "let out anything from firewall host itself"
```

Error at bootup (or deactivate the interface and activate it again)
```
opnsense: /usr/local/etc/rc.bootup: New alert found: There were error(s) loading the rules: no IP address found for ovpnc2:network - The line in question reads [0]
```

PHP-Error in Reporter:
```
[22-Feb-2018 16:09:55 Europe/Berlin] PHP Warning:  A non-numeric value encountered in /usr/local/etc/inc/filter.inc on line 473
```

**Comment by kingfisher77** [comment]
If we don't assign interfaces and just use the group interface OpenVPN Interface at Firewall>NAT>Outbound then the the connection are not stable. pings to hosts resolve in a second and the next second it does not resolve.

**Comment by AdSchellevis** [comment]
It's not very logical that it results in the same error when "Disable force gateway" is checked, there is no "ovpnc2:network" in that ruleset....  I'm not sure I can help you further with this.

**Comment by kingfisher77** [comment]
It seems the way we followed for achieving our goals was not the best one. We learned a bit more about nat and set up outbound rules which seem to work perfectly now.

Nevertheless is the possibilty to assign an vpn interface without an fixed ip address which results in an error (3 minutes waiting at boot time) possibly a weakness of the architecture.

Now we are happy again with a properly set up system. Better then before! :)
